compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/FirHelpers.kt

```diff
@@ -321,3 +323,165 @@ val FirValueParameter.hasValOrVar: Boolean
+                            && (expClassId !in StandardClassIds.primitiveTypesAndString || receiverClassId !in StandardClassIds.primitiveTypesAndString)
+                        )
+                            return ConstAndAnnotationErrorTypes.NOT_CONST
+
+                        checkConstInitializerAndAnnotationArguments(exp, session)?.let { return it }
+                    }
+                }
+                else -> {
+                    if (expression.arguments.isNotEmpty() || calleeReference !is FirResolvedNamedReference) {
+                        return ConstAndAnnotationErrorTypes.NOT_CONST
+                    }
+                    val symbol = calleeReference.resolvedSymbol as? FirCallableSymbol
+                    if (calleeReference.name == OperatorNameConventions.TO_STRING ||
+                        calleeReference.name in CONVERSION_NAMES && symbol?.callableId?.packageName?.asString() == "kotlin"
+                    ) {
+                        return checkConstInitializerAndAnnotationArguments(expression.dispatchReceiver, session)
+                    }
+                    return ConstAndAnnotationErrorTypes.NOT_CONST
+                }
+            }
+        }
+        expression is FirQualifiedAccessExpression -> {
+
+            when {
+                (expressionSymbol as FirProperty).isLocal || expressionSymbol.symbol.callableId.className?.isRoot == false ->
+                    return ConstAndAnnotationErrorTypes.NOT_CONST
+                expression.typeRef.coneType.classId == StandardClassIds.KClass ->
+                    return ConstAndAnnotationErrorTypes.NOT_KCLASS_LITERAL
+
+                //TODO: UNRESOLVED REFERENCE
+                expression.dispatchReceiver is FirThisReceiverExpression ->
+                    return null
+            }
+
+            return when ((expressionSymbol as? FirProperty)?.initializer) {
+                is FirConstExpression<*> -> {
+                    if ((expressionSymbol as? FirVariable)?.isVal == true)
+                        ConstAndAnnotationErrorTypes.NOT_CONST_VAL_IN_CONST_EXPRESSION
+                    else
+                        ConstAndAnnotationErrorTypes.NOT_CONST
+                }
+                is FirGetClassCall ->
+                    ConstAndAnnotationErrorTypes.NOT_KCLASS_LITERAL
+                else ->
+                    ConstAndAnnotationErrorTypes.NOT_CONST
+            }
+        }
+        else ->
+            return ConstAndAnnotationErrorTypes.NOT_CONST
+    }
+    return null
+}
+
+private fun FirTypedDeclaration?.getReferencedClass(session: FirSession): FirSymbolOwner<*>? =
+    this?.returnTypeRef
+        ?.coneTypeSafe<ConeLookupTagBasedType>()
+        ?.lookupTag
+        ?.toSymbol(session)
+        ?.fir
+
+private val CONVERSION_NAMES = listOf(
+    "toInt", "toLong", "toShort", "toByte", "toFloat", "toDouble", "toChar", "toBoolean"
+).mapTo(hashSetOf()) { Name.identifier(it) }
+
+enum class ConstAndAnnotationErrorTypes {
```

Comment:
I suggest a name `ConstantArgumentKind` here. May be it's better to add `CONST` kind and use this type as not null.

Reply:
Also, it's better to make the class `internal`